exchanges/kline/kline.go

```diff
@@ -507,6 +483,12 @@ func (h *IntervalRangeHolder) SetHasDataFromCandles(incoming []Candle) {
 	intervals:
 		for y := range h.Ranges[x].Intervals {
 			for z := range bucket {
+				if bucket[z].Low <= 0 && bucket[z].High <= 0 &&
```

Comment:
Due to padding adding they should be the same length, here's a suggestion below for optimization:
```diff
diff --git a/exchanges/kline/kline.go b/exchanges/kline/kline.go
index 3c8549076..d7416adb9 100644
--- a/exchanges/kline/kline.go
+++ b/exchanges/kline/kline.go
@@ -476,30 +476,33 @@ func (k *Item) GetClosePriceAtTime(t time.Time) (float64, error) {

 // SetHasDataFromCandles will calculate whether there is data in each candle
 // allowing any missing data from an API request to be highlighted
-func (h *IntervalRangeHolder) SetHasDataFromCandles(incoming []Candle) {
-       bucket := make([]Candle, len(incoming))
-       copy(bucket, incoming)
+func (h *IntervalRangeHolder) SetHasDataFromCandles(incoming []Candle) error {
        for x := range h.Ranges {
-       intervals:
                for y := range h.Ranges[x].Intervals {
-                       for z := range bucket {
-                               if bucket[z].Low <= 0 && bucket[z].High <= 0 &&
-                                       bucket[z].Close <= 0 && bucket[z].Open <= 0 &&
-                                       bucket[z].Volume <= 0 {
-                                       // function AddPadding has already been called and filled entries with empty entries
-                                       continue
-                               }
-                               cu := bucket[z].Time.Unix()
-                               if cu >= h.Ranges[x].Intervals[y].Start.Ticks &&
-                                       cu < h.Ranges[x].Intervals[y].End.Ticks {
-                                       h.Ranges[x].Intervals[y].HasData = true
-                                       bucket = bucket[z+1:]
-                                       continue intervals
-                               }
+                       offset := x + y
+                       if offset >= len(incoming) {
+                               return errors.New("holder to incoming candle length mismatch")
+                       }
+
+                       // NOTE: Can use comparative (==) used for optimization and times *should* be
+                       // truncated to intervals and stripped of monotonic clock. But it will break a few things.
+                       if h.Ranges[x].Intervals[y].Start.Time.Equal(incoming[offset].Time) {
+                               return errors.New("holder to incoming candle time clock misalignment")
+                       }
+
+                       if incoming[offset].Low <= 0 && incoming[offset].High <= 0 &&
+                               incoming[offset].Close <= 0 && incoming[offset].Open <= 0 &&
+                               incoming[offset].Volume <= 0 {
+                               // function AddPadding has already been called and filled
+                               // entries with empty entries.
+                               h.Ranges[x].Intervals[y].HasData = false
+                               continue
+                       } else {
+                               h.Ranges[x].Intervals[y].HasData = true
                        }
-                       h.Ranges[x].Intervals[y].HasData = false
                }
        }
+       return nil
 }
```